application/apps/text_grep/src/wc_1_ahocorasick.rs

```diff
@@ -0,0 +1,165 @@
+use std::sync::{mpsc, Arc};
+use std::thread;
+use tokio_util::sync::CancellationToken;
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::time::Instant;
+
+#[derive(Debug)]
+pub struct SearchResult {
+  pub file_path: String,
+  pub pattern_counts: HashMap<String, usize>,
+  pub error_message: Option<String>, // Include an optional error message field
+}
+
+pub struct TextGrep;
+
+impl TextGrep {
+  pub fn new() -> Self {
+    TextGrep
+  }
+
+  pub async fn count_occurrences(
+    &self,
+    patterns: Vec<&str>,
+    file_paths: Vec<&str>,
+    chunk_size: usize,
+    cancel_token: CancellationToken,
+  ) -> Result<Vec<SearchResult>, String> {
+    let mut results = Vec::new();
+
+    let cancel_token_clone = cancel_token.clone();
+
+    // Create a channel for sending search results from threads to the main thread
+    let (sender, receiver) = mpsc::channel();
+    let (error_sender, error_receiver) = mpsc::channel();
+
+
+    let patterns_arc: Vec<_> = patterns.iter().map(|&p| Arc::from(p)).collect();
+    let file_paths_arc: Vec<_> = file_paths.iter().map(|&fp| Arc::from(fp)).collect();
+
+    // spawn a thread for each file to read it concurrently
+    let thread_handles: Vec<_> = file_paths_arc
+      .iter()
+      .map(|file_path| {
+        let patterns = patterns_arc.clone();
+        let sender = sender.clone();
+        let error_sender_clone = error_sender.clone();
+        let cancel_token = cancel_token_clone.clone();
+        let file_path = Arc::clone(file_path);
+        thread::spawn(move || {
+          if let Err(err) = process_file(&file_path, &patterns, chunk_size, &cancel_token, &sender) {
+            if error_sender_clone.send(err.to_string()).is_err() {
+              eprintln!("Error sending error message through channel");
+            }
+          }
+        })
+      })
+      .collect();
+
+    eprintln!("Reached till before join of handles");
+    // Wait for all threads to finish
+    for handle in thread_handles {
+      handle.join().unwrap();
+    }
+
+    eprintln!("Reached till after join of handles");
+
+    while let Ok(err_msg) = error_receiver.try_recv() {
+      eprintln!("Error processing file: {:?}", err_msg);
+      // Push a SearchResult with the error message
+      results.push(SearchResult {
+        file_path: "".to_string(),
+        pattern_counts: HashMap::new(),
+        error_message: Some(err_msg.to_string()),
+      });
+    }
+
+    // Loop to receive search results
+    while let Ok(search_result) = receiver.try_recv() {
+      results.push(search_result?);
+    }
+
+    Ok(results)
+  }
+}
+
+fn process_file(
+  file_path: &Arc<str>,
+  patterns: &[Arc<str>],
+  chunk_size: usize,
+  cancel_token: &CancellationToken,
+  sender: &mpsc::Sender<Result<SearchResult, String>>,
+) -> Result<(), String> {
```

Comment:
Crate should have own Error enum. Please take a look, for example, to crate `thiserror`, which helps to create it. See examples in other places of chipmunk's solution.